src/pyqasm/visitor.py

```diff
@@ -2653,6 +2730,157 @@ def _visit_box_statement(self, statement: qasm3_ast.Box) -> list[qasm3_ast.State
+        """Visit a calibration statement element.
+
+        Args:
+            statement (qasm3_ast.CalibrationStatement): The calibration statement to visit.
+
+        Returns:
+            None
+        """
+        # TODO: Implement calibration statement handling
+        if not self._consolidate_qubits:
+            raise_qasm3_error(
+                "calibration statements requires qubit register consolidation",
+                error_node=statement,
+                span=statement.span,
+            )
+
+        self._scope_manager.push_context(Context.PULSE)
+        self._scope_manager.push_scope({})
+
+        try:
+            block_body = parse_openpulse(statement.body, in_defcal=False, permissive=False)
+        except OpenPulseParsingError as err:
+            _error_line = re.sub(
+                r"line (\d+)",
+                lambda m: f"line {statement.span.start_line + int(m.group(1)) - 1}",  # type: ignore
+                str(err),
+            )
+            raise ValidationError(f"Failed to parse OpenPulse string: {_error_line}") from err
+
+        result = []
+        calibration_stmts: list[openpulse_ast.Statement] = block_body.body
+        self._consolidate_qubits = True
+        for stmt in calibration_stmts:
+            stmt.span.start_line += statement.span.start_line  # type: ignore
+
+        result.extend(self._pulse_visitor.visit_basic_block(calibration_stmts, is_def_cal=False))
+        if len(result) == 0:
+            return []
+
+        self._scope_manager.pop_scope()
+        self._scope_manager.restore_context()
+
+        if self._check_only:
+            return [statement]
+
+        old_stdout = sys.stdout
+        captured_output = StringIO()
+        sys.stdout = captured_output
+
+        try:
+            body_str = "".join([pulse_dumps(stmt) for stmt in result])
+            body_str = re.sub(r"\n(?![\s])", "\n ", body_str)
+            body_str = re.sub(r"\n +$", "\n", body_str)
+            lines = body_str.splitlines(keepends=True)
+            for i, line in enumerate(lines):
+                if line.strip() != "":
+                    if not line.startswith(" "):
+                        lines[i] = " " + line
+                    break
+            body_str = "".join(lines)
+            _ = captured_output.getvalue()
+        finally:
+            # Restore stdout
+            sys.stdout = old_stdout
+
+        statement.body = "\n" + body_str
+
+        return [statement]
+
+    def _visit_calibration_grammar_declaration(
+        self, statement: qasm3_ast.CalibrationGrammarDeclaration
+    ) -> list[qasm3_ast.Statement]:
+        """Visit a calibration grammar declaration element.
+
+        Args:
+            statement (qasm3_ast.CalibrationGrammarDeclaration): The calibration grammar declaration
+
+        Returns:
+            None
+        """
+        # TODO: Implement calibration grammar declaration handling
+        if statement.name == "openpulse":
```

Comment:
We should also raise error if any other `defcalgrammar` is mentioned - or is that a syntax error?